src/poetry/console/commands/build.py

```diff
@@ -47,49 +59,120 @@ class BuildCommand(EnvCommand):
         "poetry.core.masonry.builders.wheel",
     ]
 
-    def _build(
-        self,
-        fmt: str,
-        executable: str | Path | None = None,
-        *,
-        target_dir: Path | None = None,
-    ) -> None:
-        from poetry.masonry.builders import BUILD_FORMATS
+    def _requested_formats(self) -> list[Literal["sdist", "wheel"]]:
+        fmt = self.option("format") or "all"
 
         if fmt in BUILD_FORMATS:
-            builders = [BUILD_FORMATS[fmt]]
+            formats = [fmt]
         elif fmt == "all":
-            builders = list(BUILD_FORMATS.values())
+            formats = BUILD_FORMATS
         else:
             raise ValueError(f"Invalid format: {fmt}")
 
+        return formats  # type: ignore[return-value]  # TODO
+
+    def _config_settings(self) -> dict[str, str]:
+        config_settings = {}
+
         if local_version_label := self.option("local-version"):
-            self.poetry.package.version = self.poetry.package.version.replace(
-                local=local_version_label
+            self.line_error(
+                f"<warning>`<fg=yellow;options=bold>--local-version</>` is deprecated."
+                f" Use `<fg=yellow;options=bold>--config-settings local-version={local_version_label}</>`"
+                f" instead.</warning>"
             )
+            config_settings["local-version"] = local_version_label
+
+        for config_setting in self.option("config-settings"):
+            key, _, value = config_setting.partition("=")
+            config_settings[key] = value
 
-        for builder in builders:
-            builder(self.poetry, executable=executable).build(target_dir)
+        return config_settings
+
+    def _build(
+        self,
+        fmt: Literal["wheel", "sdist"],
+        executable: Path,
+        target_dir: Path,
+        config_settings: dict[str, Any],
+    ) -> None:
+        from poetry.masonry.builders import BUILD_FORMATS
+
+        if fmt not in BUILD_FORMATS:
+            raise ValueError(f"Invalid format: {fmt}")
+
+        builder = BUILD_FORMATS[fmt]
+
+        builder(
+            self.poetry,
+            executable=executable,
+            config_settings=config_settings,
+        ).build(target_dir)
+
+    def _isolated_build(
+        self,
+        fmt: Literal["wheel", "sdist"],
+        executable: Path,
+        target_dir: Path,
+        config_settings: dict[str, Any],
+    ) -> None:
+        with isolated_builder(
+            source=self.poetry.file.path.parent,
+            distribution=fmt,
+            python_executable=executable,
+        ) as builder:
+            builder.build(fmt, target_dir, config_settings=config_settings)
+
+    def _requires_isolated_build(self) -> bool:
+        from poetry.core.constraints.version import Version
+
+        from poetry.utils._compat import metadata
+
+        if len(self.poetry.build_system_dependencies) > 1:
+            return True
```

Comment:
We should probably allow this as well, maybe switch logic to checking if all requirements have allowed versions in the installed repo. 

My thinking is that, users who use build scripts today might have "additional" setups in their venv or  have build dependencies locked via groups. To allow for least surprises, we could create a deprecation warning of that scenario (if we want to follow a deprecation cycle), or alternatively provide a flag that allows either an opt-out or in.

Reply:
You are correct. Forgot about #9227 :laughing:

Reply:
I've now implemented a variant to check if all build dependencies are installed in the current env and if use the env. But I'm will revert this change in a minute, because every build backend can provide a list of additional dependencies. We already respect this in the `isolated_builder()` context manager.

So the only situation in which we can be relativ sure to use the current env is, when no build script is defined, `poetry-core` is the only build dependency, the installed `poetry-core` version is within the given version range and the no source_type is given (so it is e.g. a git dependency).